### ISSUE TITLE
Feature to paste images directly into the textarea without needing image dialog

### ISSUE THREAD

**Issue by GaurangTandon** [soscripted/sox#338]
Similar to what GitHub does. If you copy an image into your clipboard and paste it into the GitHub textbox, it:

1. automatically detects that the data being pasted is an image
2. contacts the respective image server and uploads the image there
3. when the upload is complete, returns the generated URL and embeds the image as `![alt text](URL)`

I will try and see if this is possible with the SE Imgur API, and if it is, will submit a pull request with this attempted new feature. Hopefully it will be short enough, and be a useful addition to SOX :)

**Comment by GaurangTandon** [comment]
This [SO question](https://stackoverflow.com/questions/8659808/how-does-http-file-upload-work?) is supposed to help, but with my little AJAX experience, I can't figure it out :/ @shu8 do you have some experience uploading images through AJAX?

**Comment by shu8** [comment]
@GaurangTandon sure I'll have a look! I've done something very similar in another project so hopefully it won't be too hard to make it work for SOX!

**Comment by GaurangTandon** [comment]
Great! I did try this myself first though, and interestingly got the 418 error! :O 

The code is pretty short, and I am unable to figure out what's wrong (my code matches well with what I found in those links above)

```
(function() {
    'use strict';

    window.addEventListener("paste", function(event){
        var node = event.target,
            isTextarea = node.tagName === "TEXTAREA",
            clipboardData = event.clipboardData,
            formData = new FormData(),
            req = new XMLHttpRequest(),
            reader = new FileReader(),
            items = clipboardData && clipboardData.items,
            blob = null;

        if(!items) return true;

        for(var i = 0, len = items.length; i < len; i++)
            if(/^image/.test(items[i].type)){
                blob = items[i].getAsFile();
                break;
            }

        if(isTextarea && blob){
            reader.onload = function(){
                formData.append("file", this.result);

                req.open("POST", "https://stackoverflow.com/upload/image?https=true");
                req.setRequestHeader("accept", "*/*");
                req.setRequestHeader("content-type", "multipart/form-data");
                req.send(formData);

                debugger;

                req.onload = function(e){
                    debugger;
                };
            };

            reader.readAsDataURL(blob);
        }
    });
})();
```

I'll again have a look later.

**Comment by GaurangTandon** [comment]
> Also, something I didn't get whilst I was doing it is why can there be more than one item?

Yeah, I noticed a few times the `items` array was set to `["image/png", "Files"]`. Not sure why though :/

> I couldn't get it to work for the Stack Imgur API

Hmm, I am not sure if this is a good idea. Stack gives exclusive support to the `i.stack` domain, taking regular backups of the images, and ensuring their eternal permanence. While I agree that normal imgur images shouldn't vanish either, the probability of the latter is higher. Moreover, I'd personally want to replicate the default functionality as far as possible :)

> to get the fkey I think it is in the #fkey element's value on every page.

Nice find! I inserted a simple `formData.append("CSRF", document.getElementById("fkey").value);`, but that didn't still work though :/

**Comment by shu8** [comment]
@GaurangTandon I remember seeing a stackapps question on uploading images to the SE imgur account. I'll try finding it!

**Comment by shu8** [comment]
@GaurangTandon Found it! [Is there an API to upload images to SE's imgur installation?](https://stackapps.com/q/2664)

**Comment by shu8** [comment]
Sure, I'll give it another try soon!

**Comment by shu8** [comment]
@GaurangTandon I think I got it to work now in [dev 2.2.5](https://github.com/soscripted/sox/raw/dev/sox.user.js) -- can you confirm?

the URL extraction is reallllllllly messy, so do you mind having a look to see if you can clean it up? 😆

**Comment by GaurangTandon** [comment]
Okay, first off, I spent five minutes debugging why the paste event doesn't fire when I realized the setting "Paste images directly into the post editor without the image dialog" is **unchecked** by default!! :O Please check it like all others.

Then, a couple more I even converted the first few lines to:

    document.addEventListener('paste', function(e) {
        var node = e.target,
             isTextarea = node && node.tagName === "TEXTAREA";
        debugger;

but it doesn't work. The debugger doesn't fire. That's because `clipboard.items` isn't what you think it is always. Here's an example:

![image](https://user-images.githubusercontent.com/6308683/42070382-ae6423dc-7b73-11e8-8ca1-961e0424b0e4.png)

And since we need the `.getAsFile()` method, we would need the loop (that was there in my original code) in here.

Also, `e.originalEvent` errors in Chrome 67.

----

Still, after all that, I get an error :(

![image](https://user-images.githubusercontent.com/6308683/42070542-8986567e-7b74-11e8-8285-3d08b627c9b7.png)

I've made a commit with fixes for all of the stuff above, but I don't know how to fix the xhr :/

**Comment by shu8** [comment]
@GaurangTandon cool thanks! I still don't understand why the whole multiple items happens, but your fix seems perfect!

The POST error was to do with me being an idiot and only testing on Meta, so I forgot to change the URL to the `/upload/image` path of the *current* site!

Hopefully it should work now in [dev v2.2.6](https://github.com/soscripted/sox/raw/dev/sox.user.js)!

**Comment by shu8** [comment]
@GaurangTandon thanks! nice touch with the pre-selection too! (and using just `'/upload/image?https=true'` is pure genius 😆).

You're right. In the past, we've never really done anything but I've been thinking recently to add a new 'added_in_version' property to the `sox.features.info.json` file, and with new releases, the script would show a small popup to users showing new features for the version -- thoughts?

**Comment by GaurangTandon** [comment]
> nice touch with the pre-selection too! (and using just '/upload/image?https=true' is pure genius 😆).

I just found that literally a day ago while google searching :P 

>  I've been thinking recently to add a new 'added_in_version' property to the sox.features.info.json

that would be too much of a burden. You could simply keep updating the contents of the popup manually. There would be two types of users: a) old and b) new. Is there a way to detect which ones are old (and are only updating their userscript) and which ones are new (and are actually installing the userscript)? If that is possible then we could only show the popup to the old users and we're done :D

**Comment by shu8** [comment]
Hmm, I'll look into that more. If there's not an out-of-the-box way, a
simple solution would be to store a GM setting `lastVersion` and compare it
to the current version (which is easily obtainable), setting it to the
current version every time it turns out to be different.

**Comment by GaurangTandon** [comment]
Not sure if that would work, considering that currently both old and new
users don't have that `lastVersion` property.

If there's not an out of the box available to us, our best method would be
to first ship the lastVersion property and only after that ship any new
feature.

**Comment by shu8** [comment]
@GaurangTandon fair enough, but I'm sure new users wouldn't mind a 'new in this version' popup anyway seeing as it is technically a new version for them 😆 

I'll try doing something with this soon.

**Comment by GaurangTandon** [comment]
I just noticed that this has already been implemented in a much cooler manner by [Tomas Zato on StackApps](https://stackapps.com/questions/4999/stackpaste-paste-images-into-the-text-form-instantly). Considering that it's 986 lines worth of code, I'm sure it can't be merged into SOX. So, it'd rather be separately on its own.

<Thought I'd just mention.>
